src/J2N/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,2125 @@
+        public void Insert(int index, [AllowNull] TKey key, TValue value)
+        {
+            if ((uint)index > (uint)_count)
+            {
+                ThrowIndexArgumentOutOfRange();
+            }
+
+            // J2N: allow null keys
+            //ThrowIfNull(key, ExceptionArgument.key);
+
+            TryInsert(index, key, value, InsertionBehavior.ThrowOnExisting);
+        }
+
+        /// <summary>Removes the value with the specified key from the <see cref="OrderedDictionary{TKey, TValue}"/>.</summary>
+        /// <param name="key">The key of the element to remove.</param>
+        /// <returns></returns>
+        public bool Remove([AllowNull] TKey key)
+        {
+            // The overload Remove(TKey key, out TValue value) is a copy of this method with one additional
+            // statement to copy the value for entry being removed into the output parameter.
+            // Code has been intentionally duplicated for performance reasons.
+
+            // J2N: allow null keys
+            //ThrowIfNull(key, ExceptionArgument.key);
+
+            // Find the key.
+            int index = IndexOf(key);
+            if (index >= 0)
+            {
+                // It exists. Remove it.
+                Debug.Assert(_entries is not null);
+
+                RemoveAt(index);
+
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>Removes the value with the specified key from the <see cref="OrderedDictionary{TKey, TValue}"/> and copies the element to the value parameter.</summary>
+        /// <param name="key">The key of the element to remove.</param>
+        /// <param name="value">The removed element.</param>
+        /// <returns>true if the element is successfully found and removed; otherwise, false.</returns>
+        public bool Remove([AllowNull] TKey key, [MaybeNullWhen(false)] out TValue value)
+        {
+            // J2N: allow null keys
+            //ThrowIfNull(key, ExceptionArgument.key);
+
+            // Find the key.
+            int index = IndexOf(key);
+            if (index >= 0)
+            {
+                // It exists. Remove it.
+                Debug.Assert(_entries is not null);
+
+                value = _entries![index].Value; // [!]: asserted above
+                RemoveAt(index);
+
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        /// <summary>Removes the key/value pair at the specified index.</summary>
+        /// <param name="index">The zero-based index of the item to remove.</param>
+        public void RemoveAt(int index)
+        {
+            int count = _count;
+            if ((uint)index >= (uint)count)
+            {
+                ThrowIndexArgumentOutOfRange();
+            }
+
+            // Remove from the associated bucket chain the entry that lives at the specified index.
+            RemoveEntryFromBucket(index);
```

Comment:
I noticed a bug that we might want to report upstream. In `Dictionary<TKey, TValue>`, when a bucket is cleared it also clears the values if there are any reference types involved so the GC can collect them. 

https://github.com/dotnet/runtime/blob/v8.0.15/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1010-L1018

But it appears that `OrderedDictionary<TKey, TValue>` will hold onto these references until the bucket is reused or the collection is GC'd. Since Java only uses reference types, this will be a problem we will run into frequently.

Reply:
I initially agreed with your assessment that that's how the code seems, however, when I wrote an app to try to reproduce this with large arrays as the values before filing the bug upstream, I found it was actually properly GC'ing it in both the J2N and .NET 9 versions. After reading the code again, I figured out where that was happening: at the `entries![--_count] = default;` line below this one after it shifts the entries. So I'll remove my "fix" and leave it be. It is possible when this is converted to use "holes," however, that you will need to take a different approach.